src/Chapter21/Listing21.10.CancelingParallelLoop.cs

```diff
@@ -35,52 +36,58 @@ public static async Task Main()
             CancellationTokenSource cts =
                 new CancellationTokenSource();
 
-            Console.WriteLine("Push ENTER to Exit.");
-
-            // Use Task.Factory.StartNew<string>() for
-            // TPL prior to .NET 4.5
             Task task = Task.Run(() =>
             {
                 data = ParallelEncrypt(data, cts.Token);
             }, cts.Token);
 
-            // Wait for the user's input
-            Console.Read();
+            Console.WriteLine("Push ENTER to Exit.");
+            Task<int> cancelTask = ConsoleReadAsync(cts.Token);
 
-            if (!task.IsCompleted)
+            try
             {
+                Task.WaitAny(task, cancelTask);
+                // Cancel which ever task has not finished.
                 cts.Cancel();
-                try { task.Wait(); }
-                catch (AggregateException exception)
+                await task;
+
+                Console.ForegroundColor = ConsoleColor.Green;
+                Console.WriteLine("\nCompleted successfully");
+
+            }
+            catch(OperationCanceledException taskCanceledException)
+            {
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.WriteLine(
+                    $"\nCancelled: { taskCanceledException.Message }");
+            }
+            Console.ForegroundColor = originalColor;
+        }
+
+        private static async Task<int> ConsoleReadAsync(
+            CancellationToken cancellationToken = default)
+        {
+            int result = 0;
+            await Task.Run(async () =>
+            {
+                int maxDelay = 1025;
+                int delay = 0;
+                while (!cancellationToken.IsCancellationRequested)
                 {
-                    Console.ForegroundColor = ConsoleColor.Red;
-                    TaskCanceledException taskCanceledException =
-                        (TaskCanceledException)exception.Flatten()
-                        .InnerExceptions.FirstOrDefault(
-                        innerException =>
-                            innerException.GetType() ==
-                            typeof(TaskCanceledException));
-                    if(taskCanceledException != null){
-                        Console.WriteLine($@"Cancelled: { 
-                            taskCanceledException.Message }");
+                    if (Console.KeyAvailable)
+                    {
+                        result = Console.Read();
+                        break;
```

Comment:
The output above says to press enter to exit, but this will break on all key presses.